### DISCUSSION TITLE
How run it locally with our own GPU?

### DISCUSSION THREAD

**Discussion by ZeroCool22** [lkwq007/stablediffusion-infinity#45] [Q&A]
git clone --recurse-submodules https://github.com/lkwq007/stablediffusion-infinity
cd stablediffusion-infinity
# run jupyter notebook or jupyter lab here

**Don't know what to do in the last step to run it Locally:** 

# run jupyter notebook or jupyter lab here

**Comment by urbanscribe** [comment]
he means run the notebook

`jupiter notebook`

is the command
you need to install it pip/conda

**Comment by ZeroCool22** [comment]
> pip/conda

Please tell me the exact commands i need to use:

![Screenshot_1](https://user-images.githubusercontent.com/13344308/190279114-7586da93-a736-4663-8743-ece619cb0ae9.jpg)

Also, i have Anaconda/Conda installed already:

![Screenshot_2](https://user-images.githubusercontent.com/13344308/190279220-04c43409-f2d9-4a0f-988c-1bd2da027ca6.jpg)

Thx in advance!

**Comment by ZeroCool22** [comment]
> he means run the notebook
> 
> `jupiter notebook`
> 
> is the command you need to install it pip/conda

![Screenshot_4](https://user-images.githubusercontent.com/13344308/190280728-83c03f11-7612-4778-b2e3-d23e5e0b9c0f.jpg)

**Comment by ZeroCool22** [comment]
well, it doesn't run...

**Comment by DenkingOfficial** [comment]
Just type those commands one by one

```
conda create -n sdinf python=3.10 -y
conda activate sdinf
pip install notebook
jupyter notebook
```

**Comment by ZeroCool22** [comment]
> Just type those commands one by one
> 
> ```
> conda create -n sdinf python=3.10 -y
> conda activate sdinf
> pip install notebook
> jupyter notebook
> ```

Big thx, i'm in.

![Screenshot_6](https://user-images.githubusercontent.com/13344308/190282661-3c684848-5d9a-44bb-9439-133f555f46c4.jpg)

Now, how i get to the Canvas to start Outpainting?

**Comment by ZeroCool22** [comment]
![cxccccccc](https://user-images.githubusercontent.com/13344308/190283570-89eda9a2-4915-45f6-b1d2-4bba6187a848.png)

Errors in every cell i try to RUN...

**Comment by lkwq007** [comment]
For Windows, you may need to install anaconda or miniconda at first and then:
```
conda create -n sd-inf python=3.10
conda activate sd-inf
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
conda install scipy
conda install -c conda-forge jupyterlab 
conda install -c conda-forge ipycanvas
conda install -c conda-forge diffusers transformers ftfy
pip install opencv-python
git clone https://github.com/lkwq007/stablediffusion-infinity
cd stablediffusion-infinity
huggingface-cli login
jupyter lab
# and open stablediffusion_infinity_windows.ipynb
```
Note that the notebook has not been tested on Windows, so I am not sure whether it would work well.

**Comment by Kim2091** [comment]
@lkwq007 Hi, i'm not the OP of this thread, but I'm attempting to utilize this on Windows. All of this seems to have worked, but I needed to use `conda install opencv` instead of `pip install opencv-python`. 

Sadly i'm running into this issue in the upload image cell in the `stablediffusion_infinity_windows.ipynb`:

> 
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> Cell In [7], line 14, in start_func(btn)
>      12 uploader_output.clear_output()
>      13 if len(uploader.value)>0:
> ---> 14     keys=list(uploader.value.keys())
>      15     val=uploader.value[keys[-1]]["content"]
>      16     pil=Image.open(io.BytesIO(val))
> 
> AttributeError: 'tuple' object has no attribute 'keys'

**Comment by urbanscribe** [comment]
Kim - running into same "tuple" error AT TIMES
the uploader is confusing, you can select the image but when confirming says no image uploaded 
have tried to crop the images to ingest to 512/512 in case resizer was a problem

if useful
- instruction don't work on M1 macs (for obvious ARM reasons).  Windows works
- make sure you are using a token from an account on hugging face where you approved the Terms and conditions

Help on how to upload an image which is placed in the working area would be appreciated

**Comment by Kim2091** [comment]
Thank you for the reply. I am on Windows, and I've now verified that I'm logged in. Sadly it's still yielding the same error. Inpainting itself actually does function, it's just uploading an image that is broken as you specified.

**Comment by lkwq007** [comment]
> @lkwq007 Hi, i'm not the OP of this thread, but I'm attempting to utilize this on Windows. All of this seems to have worked, but I needed to use `conda install opencv` instead of `pip install opencv-python`.
> 
> Sadly i'm running into this issue in the upload image cell in the `stablediffusion_infinity_windows.ipynb`:
> 
> > AttributeError                            Traceback (most recent call last)
> > Cell In [7], line 14, in start_func(btn)
> > 12 uploader_output.clear_output()
> > 13 if len(uploader.value)>0:
> > ---> 14     keys=list(uploader.value.keys())
> > 15     val=uploader.value[keys[-1]]["content"]
> > 16     pil=Image.open(io.BytesIO(val))
> > AttributeError: 'tuple' object has no attribute 'keys'

Seems to be `ipywidgets` issue, how about `conda install -c conda-forge ipywidgets=7.7.1 nodejs`?

**Comment by Kim2091** [comment]
> conda install -c conda-forge ipywidgets=7.7.1 nodejs

Thank you, that fixed the issue with uploading images. Unfortunately a new error has occurred in the following cell: 

> ---------------------------------------------------------------------------
> NameError                                 Traceback (most recent call last)
> Cell In [6], line 4
>       2 upload_button.disabled=True
>       3 display_lst=base.display()
> ----> 4 display_lst.insert(-1,widgets.HBox([resize_check,step_input,guidance_input,strength_slider]))
>       5 for item in display_lst:
>       6     display(item)
> 
> NameError: name 'resize_check' is not defined

EDIT: Never mind, my apologies. A cell failed to execute and I overlooked it. Thank you for the help!

**Comment by devilkkw** [comment]
but how to use local pretrained models? i have no account to hugginface, is possiblo to run offline with downloaded pretrained models? (like automatic111 if you know)

**Comment by urbanscribe** [comment]
you can open a free account on HF.  you'll need an access token

**Comment by spinferno** [comment]
> conda install -c conda-forge ipywidgets=7.7.1 nodejs

Thanks for the command!
after I run that, start again and get this error in the 'upload an image' and 'have fun here' cells:
![image](https://user-images.githubusercontent.com/12758205/190543499-385d3add-7635-439d-95c6-0c71f9eb0d92.png)
what can I do? thanks for your time!!

**Comment by Kim2091** [comment]
> after I run that, start again and get this error in the 'upload an image' and 'have fun here' cells:

@spinferno Try installing jupyterlab_widgets 1.1.1, that seems to work for me

**Comment by spinferno** [comment]
> > after I run that, start again and get this error in the 'upload an image' and 'have fun here' cells:
> 
> @spinferno Try installing jupyterlab_widgets 1.1.1, that seems to work for me

thanks so much for your help!
as per the latest commit:
https://github.com/lkwq007/stablediffusion-infinity/commit/7a2a1d985fdcd023ac69a6b2240ee61d42d29a47

I ran these:

`conda install -c conda-forge ipywidgets=7.7.1 nodejs`
`conda install -c conda-forge jupyterlab_widgets=1.1.1 nodejs`
`conda install -c conda-forge widgetsnbextension=3.6.1 nodejs`

however same exception is thrown, sadly.

**Comment by Kim2091** [comment]
🙂 You may want to make a fresh clone of the repo and try a new environment from scratch. I ran into similar issues, doing this seems to fix it. Maybe having the older dependencies and running it at all somehow introduces permanent issues to that copy?

I hope you get this figured out. I made an issue #12 that documented many of the issues I ran into earlier today. You may want to check that out for any solutions.

**Comment by spinferno** [comment]
i just wiped my env and stood up everything again. somehow the exception remains and now all the UI buttons are still broken, include the generate button, how else can i fix things?
this is what I did:

_remove installation_
conda env remove -n sd-inf

_rebuild_ 
conda create -n sd-inf python=3.10 -y
conda activate sd-inf
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch -y
conda install scipy -y
conda install -c conda-forge jupyterlab -y
conda install -c conda-forge ipywidgets=7.7.1 -y
conda install -c conda-forge ipycanvas -y
conda install -c conda-forge diffusers transformers ftfy -y
pip install opencv-python
conda activate sd-inf
jupyter lab
_tested, shut down_

_fix dependencies_
conda install -c conda-forge ipywidgets=7.7.1 nodejs
conda install -c conda-forge jupyterlab_widgets=1.1.1 nodejs
conda install -c conda-forge widgetsnbextension=3.6.1 nodejs
jupyter lab
_tested, still getting runtime errors_

### runtime errors 
for the file upload widget:
[Open Browser Console for more detailed log - Double click to close this message]
Failed to load model class 'LabelModel' from module '@jupyter-widgets/controls'
Error: Module @jupyter-widgets/controls, version ^1.5.0 is not registered, however,         2.0.0 is
    at f.loadClass (http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/134.bcbea9feb6e7c4da7530.js?v=bcbea9feb6e7c4da7530:1:74977)
    at f.loadModelClass (http://localhost:8888/lab/extensions/@jupyter-widgets/jupyterlab-manager/static/150.3e1e5adfd821b9b96340.js?v=3e1e5adfd821b9b96340:1:10729)

for the uplaod confirm button:
[Open Browser Console for more detailed log - Double click to close this message]
Failed to load model class 'HBoxModel' from module '@jupyter-widgets/controls'
Error: Module @jupyter-widgets/controls, version ^1.5.0 is not registered, however,         2.0.0 is

my conda version is:
          conda version : 4.14.0
         python version : 3.9.12.final.0
   
have I done something to my PC that is preventing me from even recreating my local conda env? what should I wipe to at least get the generate button back? I am at a loss. I have deleted my env folders. should I delete my conda package cache too?

**Comment by fabstr1** [comment]
![image](https://user-images.githubusercontent.com/72360831/210155844-0bfdf904-70fa-4a21-83e6-906fc3ca7304.png)
Hello, I have followed the instruction to download the program in this thread. But when I try to run it, I can't get any further than for the program to be stuck in the running mode, see my screenshot. I have waited for around one hour for the program to finish. Have I missed something or should it take time for the program to start?

**Comment by fabstr1** [comment]
![image](https://user-images.githubusercontent.com/72360831/210156647-f174f6a9-0aa3-4b93-991d-a36b59a1a84b.png)
Update. Now I have the error message, "fatal: destination path 'stablediffusion-infinity' already exists and is not an empty directory."
